### DISCUSSION TITLE
Question about Sirius usage

### DISCUSSION THREAD

**Discussion by phamvietquang123** [eclipse-sirius/sirius-desktop#370] [Q&A] [answered]
Hi i am a beginner using Sirius Eclipse and am using this as my metamodel 
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/91275915/61e10de4-f98e-4097-bc8c-fa35f12aa909)
In Sirius i have to create an activityedge node to connect from source node to target node. Is there an option to directly connect the source node to the target node without including an activity edge node?"

**Comment by GlennPlou** [comment]
Hi, 

I'm not sure I understand what you're trying to do: are you trying to reproduce the UML Activity metamodel with Ecore? Are you manipulating mappings in the .odesign file?

Regards,
Glenn

**Reply by phamvietquang123** [reply]
I want to manipulating mappings in the .odesign file. i have a part of a model like this 
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/91275915/8fef1068-f2e1-46a1-9449-72ad154d410c)
Currently, when I visualize it using Sirius, it is displayed like this:
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/91275915/31a26008-1458-4423-9e02-6b11ef151ac2)
However, I want to connect directly from 'Handle Order' to 'CustOrder' without going through the 'Edge' node."

**Reply by GlennPlou** [reply, marked as answer]
The "ActivityEdge" type you want to represent should be an Edge Mapping in the odesign. So you need to determine whether it's an "Element based edge" or a "Relation based edge": here, the edge shouldn't represent a relation (incoming, outgoing, source, target...) but directly a semantic element, so you need to use an "Element based edge" mapping that represents "ActivityEdge" type elements.

![image](https://github.com/eclipse-sirius/sirius-desktop/assets/33932970/c35c9eaf-69d0-4366-903b-4ba32fac161c)


I've attached a screenshot of a draft I made to help you. Note that I've included the ActivityNode mapping for source and target, but I think ActivityNode should be an abstract type and doesn't necessarily need a representation. You'll need to represent all concrete types (ForkNode, JoinNode, ...), and add their mappings to ActivityEdgeMapping's "Source Mapping" and "Target Mapping".

For the Source Finder Expression, use the self.source reference if possible, otherwise the self.getSource() method if available. Ditto for the Target Finder Expression, self.target or self.getTarget().

**Reply by phamvietquang123** [reply]
Thanks, it works as expected. I have another question: I have an AQL query: aql: self.node->reject(e | e.oclIsKindOf(agl::ControlNode)) which I use in Sirius to create nodes that are not control nodes. Initially, it works, but after some time, this error appears:
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/91275915/e2ca5064-76fa-4a29-8ec8-3e708880cc38)
However, when I use the Sirius interpreter, the query works with my model. Here is my Ecore metamodel:
[mosa.zip](https://github.com/eclipse-sirius/sirius-desktop/files/15108064/mosa.zip)

**Reply by GlennPlou** [reply]
The "`->`" operator can only be used on collections, the notation of the `node` reference may be misleading (use "nodes" in plural if it's [0...*] ?). 
When using nested EPackages (mosa > agl > ControlNode), the syntax for referring to type like agl::ControlNode may be different, in which case I don't know what the "recommended" syntax is in such cases (but if I remember correctly, it should work).

Try maybe something like `aql:self.node->excludesAll(self.node->filter(agl::ControlNode))`

Otherwise, there is a lot of aql examples here: https://eclipse.dev/acceleo/documentation/aql.html

**Reply by phamvietquang123** [reply]
it still does not work but thanks.
